perf: improve performance when spec lint [INS-3724]

CurryYangxx opened the pull request:
<!--
Please open an [Issue](https://github.com/kong/insomnia/issues/new) first to discuss new
features or non-trivial changes. Please provide as much detail as possible on the change as
possible including general description, implementation details, potential shortcomings, etc.

If this PR closes an issue, please mention "Closes #XX" where #XX is the issue number.

If this PR fixes a bug or regression, please make sure to add a test.
-->

Close https://github.com/Kong/insomnia/issues/4653 https://github.com/Kong/insomnia/issues/4241

**Move spec lint logic to web worker, can improve both opening file and editing file**

Changes:

- [x] move spectral lint login to web worker both in the Design component and router action
    - [x] add dedicate worker file
    - [x] delete all the spectral in main process
- [x] open electron nodeIntegrationInWorker option
- [x] add vite worker plugin config so we can use electron require magic in worker
- [x] wrap code-editor with memo to reduce rerender

How to test:
- [x] trigger lint multiple times and only the latest task result will be updated
- [x] use custom spectral ruleset and it it works

CurryYangxx committed:
perf: spec lint

gatzjames committed:
Move ruleset loading to main

jackkav commented:
@CurryYangxx very elegant approach. With node access in webworkers we have opened up some interesting performance options.

jackkav commented on tsconfig.base.json:
```diff
@@ -10,7 +10,7 @@
     "esModuleInterop": true,
     "lib": ["DOM", "ES2019"],
     "module": "CommonJS",
-    "moduleResolution": "node",
```
<img width="736" alt="image" src="https://github.com/Kong/insomnia/assets/3679927/d2278f12-38b6-40b8-926e-e3dab4b82a82">
curious

jackkav commented on packages/insomnia/src/ui/routes/design.tsx:
```diff
@@ -251,24 +229,71 @@ const Design: FC = () => {
-      const diagnostics = await window.main.spectralRun({
-        contents,
-        rulesetPath,
+      // @TODO: Conisderations and things we need to verify/test:
+      // - Should we spawn a new worker on every lint command or use a shared worker?
+      // - Does this function handle race conditions? e.g. if a lint finishes late does it replace another lint that ran with latest contents?
+      // - Proper error handling from the worker and here. Perhaps the worker returns a result type like Result = Diagnostics | Error and we handle it here
+      // - Don't reload the spectral ruleset on every lint command. Instead the ruleset should be re-calculated when the underlying file changes due to e.g. git pull or the filename changes
```
> Should we spawn a new worker on every lint command or use a shared worker?

it depends on the error handling story

> Does this function handle race conditions?

no if you always spawn a new one, although since it can happen on keystroke we might need to throttle the creation of workers, or resetting a shared one.

> Proper error handling from the worker and here.

the simplest way I can imagine is wrapping the worker in a promise and using abort and reject when listening on the worker error event.

> Don't reload the spectral ruleset on every lint command
agreed, but perhaps we can consider this is an optimisation for later.

CurryYangxx committed:
fix: solve todo

CurryYangxx committed:
fix: solve todo

CurryYangxx committed:
feat: open nodeIntegrationInWorker option

CurryYangxx committed:
fix: need save spectralRun ipc

CurryYangxx committed:
feat: add worker plugin

CurryYangxx commented:
> Should we spawn a new worker on every lint command or use a shared worker?

Now it will create just one worker when component first time render.And use a unique id to distinguish worker message.

> Does this function handle race conditions?

Use the unique id for every message, only the result which id match the latested id will be updated.

> Proper error handling from the worker and here.

When there is a error, it will return a rejected Promise to codemirror, so that codemirror will do nothing

> Don't reload the spectral ruleset on every lint command

already move load ruleset logic to outer function

CurryYangxx commented:
We still need to save spectralRun IPC because it is used in this action:https://github.com/Kong/insomnia/blob/cc421d1a3c041cebc059f5d421c4406c8fe75baf/packages/insomnia/src/ui/routes/actions.tsx#L736

gatzjames committed:
put the default ruleset handling in loadSpectralRuleset

gatzjames committed:
make sure we check for latest id

CurryYangxx committed:
fix: postMessage and ipc can not transfer special objects

gatzjames committed:
move logic to new place and clean up main

CurryYangxx committed:
fix race condition and cleanup code

ihexxa commented on packages/insomnia/src/ui/routes/actions.tsx:
```diff
@@ -772,7 +776,9 @@ export const generateCollectionAndTestsAction: ActionFunction = async ({ params
-  const results = (await window.main.spectralRun({ contents: apiSpec.contents, rulesetPath })).filter(isLintError);
+  const spectralRunner = new SpectralRunner();
+
+  const results = (await spectralRunner.runDiagnostics({ contents: apiSpec.contents, rulesetPath })).filter(isLintError);
```
do we also need termination here?

CurryYangxx committed:
fix: worker build

ihexxa commented on packages/insomnia/src/ui/worker/spectral.ts:
```diff
@@ -0,0 +1,65 @@
+  ruleset: oas as RulesetDefinition,
+};
+
+const loadRuleset = async (rulesetPath: string,) => {
```
nit: `(rulesetPath: string,)` the comma might be unnecessary.

ihexxa commented on packages/insomnia/src/ui/worker/spetral-run.ts:
```diff
@@ -0,0 +1,37 @@
+  public async runDiagnostics({ contents, rulesetPath }: { contents: string; rulesetPath: string }) {
+    this.taskId = ++this.taskId;
+    return new Promise<ISpectralDiagnostic[]>(resolve => {
+      this.worker.onmessage = e => {
```
nit: Probably we might also handle `postMessage(err);` and add some logs for debugging.

ihexxa reviewed (dismissed):
Discussed offline and resolved several concerns, just added some minor comments. Overall it looks good to me but let's wait a bit as there might be some comments from other reviewers.

CurryYangxx committed:
add necessary worker ternminate and more logs

CurryYangxx commented:
> Discussed offline and resolved several concerns, just added some minor comments. Overall it looks good to me but let's wait a bit as there might be some comments from other reviewers.

@ihexxa Good advice. The code has been updated

> Discussed offline and resolved several concerns, just added some minor comments. Overall it looks good to me but let's wait a bit as there might be some comments from other reviewers.

@ihexxa Good catch, update more logs and necessary termination

jackkav committed:
use fs require in worker

jackkav committed:
rename file and add reject

jackkav committed:
fix typo

jackkav committed:
remove unused console log in loader

jackkav committed:
leave note about GC

CurryYangxx committed:
Merge branch 'develop' into perf/web-worker-lint-spec
